### PR TITLE
[fix]: [DEL-3188]: Introduced an API to update/delete tags for delegate groups

### PR DESCRIPTION
**Pull request by gaurav-jain1**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)

### CODE REVIEW

File: 420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java [hunk 1]
```diff
@@ -209,4 +215,34 @@ public DelegateSetupResourceV2(DelegateSetupService delegateSetupService, Access
...
+  @PUT
+  @Path("tags/{identifier}")
+  @Timed
+  @ExceptionMetered
+  @AuthRule(permissionType = LOGGED_IN)
+  @Operation(operationId = "updateTagsForDelegateGroup", summary = "Updates tags for the Delegate Group.",
+      responses =
+      {
+        @io.swagger.v3.oas.annotations.responses.
+        ApiResponse(responseCode = "default", description = "Delegate Group details for updated group.")
```
Comment (bharatgoelharness) [inline 420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java:238]: can we change the description of swagger

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java [hunk 2]
```diff
@@ -525,4 +527,33 @@ public DelegateGroupListing listDelegateGroupDetails(
         .map(key -> (String) key.getId())
         .collect(toList());
   }
+
+  @Override
+  public DelegateGroup updateDelegateGroupTags(String accountId, String delegateName, List<String> tags) {
```
Comment (bharatgoelharness) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java:532]: delegate name -> delegate group name

File: 420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java [hunk 3]
```diff
@@ -209,4 +215,34 @@ public DelegateSetupResourceV2(DelegateSetupService delegateSetupService, Access
...
+    public List<String> getTags() {
+      return tags;
+    }
+    public void setTags(List<String> tags) {
+      this.tags = tags;
+    }
+  }
+
+  @PUT
+  @Path("tags/{identifier}")
```
Comment (bharatgoelharness) [inline 420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java:230]: change api/setup/delegate/tags/id ->api/setup/delegate/id/tags

**Review by mdmilic — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 220 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3dBqO0xS0d-s--nJ1:-Mw3dBqO0xS0d-s--nJ2:b-mc71rm) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L220)):*
> ```Java
> 
>   protected static class DelegateTags {
>     private List<String> tags;
> ```

This class should go to 920-delegate-service-beans



_Code quote:_
```Java
  protected static class DelegateTags {
    private List<String> tags;
    public List<String> getTags() {
      return tags;
    }
    public void setTags(List<String> tags) {
      this.tags = tags;
    }
  }
```
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 220 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3dYEEC63bBh0V-Jtr:-Mw3dYEEC63bBh0V-Jts:b29qde6) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L220)):*
> ```Java
> 
>   protected static class DelegateTags {
>     private List<String> tags;
> ```

make it final
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 221 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3dcIkEjHdX6LRbizV:-Mw3dcIkEjHdX6LRbizW:b-7tun5k) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L221)):*
> ```Java
>   protected static class DelegateTags {
>     private List<String> tags;
>     public List<String> getTags() {
> ```

You can use lombok annotations for getter/setter, but also you probably don't need a setter. You should make the tags field `final` and use `@Data` which would give you the right constructor and getter. 
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 233 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3eSXwAPudhiQr696X:-Mw3eSXwAPudhiQr696Y:bgnemhc) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L233)):*
> ```Java
>   @Timed
>   @ExceptionMetered
>   @AuthRule(permissionType = LOGGED_IN)
> ```

My understanding is we want this API invoked by customers, for this we should use `@ApiKeyAuthorized` authentication with correct permissions
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 534 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3fC673jFKReneffwv:-Mw3fC673jFKReneffww:bhb7l2) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L534)):*
> ```Java
>   public DelegateGroup updateDelegateGroupTags(String accountId, String delegateName, List<String> tags) {
>     Query<DelegateGroup> updateQuery = persistence.createQuery(DelegateGroup.class)
>                                            .filter(DelegateGroupKeys.accountId, accountId)
> ```

If this is going to be API for NG we need to add the owner (org/project) as filters in the query, but also isNg in general (so that you don't update a CG delegate by accident). Right now delegate group name is unique per account/isNg, but this is too strict and we want to remove that restriction, so it will be unique per accountId/orgId/projectId/isNg
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 537 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3gKeJ-JJdmghDOmQq:-Mw3gKeJ-JJdmghDOmQr:bdbk1x3) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L537)):*
> ```Java
>                                            .filter(DelegateGroupKeys.name, delegateName);
> 
>     UpdateOperations<DelegateGroup> updateOperations = persistence.createUpdateOperations(DelegateGroup.class);
> ```

In general it's a good practice to mark all your fields & variables `final` (to make them kinda immutable). I know we are not doing great job of it, but we should do better :) 
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 544 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3h1bs0gCmV1kuiwlF:-Mw3h1bs0gCmV1kuiwlG:b-61ru96) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L544)):*
> ```Java
>     delegateCache.invalidateDelegateGroupCache(accountId, updatedDelegateGroup.getUuid());
> 
>     Query<Delegate> delegatesToBeUpdated = persistence.createQuery(Delegate.class)
> ```

Do we have tags on delegate instances as well (not just groups)? Wasn't aware of that. Or why do we need this (CG?). 
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 552 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3iG_v5gheQ_mwcQtw:-Mw3iG_v5gheQ_mwcQtx:bqngr8s) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L552)):*
> ```Java
>     List<String> updatedUuid = new ArrayList<String>();
>     for (Delegate delegate : delegatesToBeUpdated.asList()) {
>       updatedUuid.add(delegate.getUuid());
> ```

You could use streams to make it a bit nicer (if you really need cache)

```delegatesToBeUpdated.asList().stream().peek(delegate -> delegateCache.get(accountId, delegate.getUuid(), true)).collect(toList());```



_Code quote:_
```Java
    List<String> updatedUuid = new ArrayList<String>();
    for (Delegate delegate : delegatesToBeUpdated.asList()) {
      updatedUuid.add(delegate.getUuid());
      delegateCache.get(accountId, delegate.getUuid(), true);
    }
```
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 553 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3j0Br144HbGHjvK5p:-Mw3j0Bs8Xv6qVNU4Tcu:bkf4lo9) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L553)):*
> ```Java
>     for (Delegate delegate : delegatesToBeUpdated.asList()) {
>       updatedUuid.add(delegate.getUuid());
>       delegateCache.get(accountId, delegate.getUuid(), true);
> ```

Not sure why we need this if you are ignoring the result?

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: 1 change requests. 10. unresolved discussions. 0 of 1 approvals obtained. 

___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 220 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3dYEEC63bBh0V-Jtr:-Mw5soJnCPDnoz7-TGHu:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L220)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

make it final
</blockquote></details>

Done.
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 220 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3dBqO0xS0d-s--nJ1:-Mw5soz3DTl9hUEXkGZb:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L220)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

This class should go to 920-delegate-service-beans
</blockquote></details>

Done.
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 221 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3dcIkEjHdX6LRbizV:-Mw5so0WDWtXgUI90jAB:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L221)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

You can use lombok annotations for getter/setter, but also you probably don't need a setter. You should make the tags field `final` and use `@Data` which would give you the right constructor and getter. 
</blockquote></details>

Done.
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 230 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mvx9PsZxiB7wpvumV4d-r1-230:-Mw5ZZ9yAzNCyYH_p8N0:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L230)):*
<details><summary><i>Previously, bharatgoelharness (Bharat Goel) wrote…</i></summary><blockquote>

change api/setup/delegate/tags/id ->api/setup/delegate/id/tags
</blockquote></details>

Done.
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 233 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3eSXwAPudhiQr696X:-Mw5ZcAS4iV7mXiRzWqh:b9pa0p6) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L233)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

My understanding is we want this API invoked by customers, for this we should use `@ApiKeyAuthorized` authentication with correct permissions
</blockquote></details>

We are writing it for both. In future UI can also use this to manipulate delegate tags.
Do we still need to add @ApiKeyAuthorized
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 532 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mvx9PsZxiB7wpvumV4e-r1-532:-Mw5_tz2-AFtgvEDmJwI:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L532)):*
<details><summary><i>Previously, bharatgoelharness (Bharat Goel) wrote…</i></summary><blockquote>

delegate name -> delegate group name
</blockquote></details>

Done.
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 534 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3fC673jFKReneffwv:-Mw5jrm1EDZQO-ImVg5d:b-gb59uu) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L534)):*
> for NG we need to add the owner (org/project) as filters in the query, but also isNg in general (so that you don't update a CG delegate by acci
I see, as of now this API will be invoked directly via customer (via script or something) so its tedious for them to send org/project ids. Let's keep this as is for now, till we make UI also to use this.
I'll add isNg in filter to avoid manipulating CG delegates.
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 544 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3h1bs0gCmV1kuiwlF:-Mw5r6KH3jqV0TetZD40:b-liyege) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L544)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Do we have tags on delegate instances as well (not just groups)? Wasn't aware of that. Or why do we need this (CG?). 
</blockquote></details>

We have tags maintained both in DelegateGroups in Delegate collection. Even I am not sure of the history behind this. But this piece is to make collections consistent across.
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 552 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3iG_v5gheQ_mwcQtw:-Mw5r_iRDLX6kW3yB1BS:b-qxfgqm) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L552)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

You could use streams to make it a bit nicer (if you really need cache)

```delegatesToBeUpdated.asList().stream().peek(delegate -> delegateCache.get(accountId, delegate.getUuid(), true)).collect(toList());```
</blockquote></details>

I am updating 'updatedUuid' here to print it as log later in this function. Not sure if streams can do that. Leaving as is.
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 553 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3j0Br144HbGHjvK5p:-Mw5sNt03AeP4AlAFS-r:bpd9mf9) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L553)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Not sure why we need this if you are ignoring the result?
</blockquote></details>

This is to update delegatecache(s), since we updated the entries in backend.

**Comment by raghuAtWings** [comment]
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 233 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3eSXwAPudhiQr696X:-MwPnaaz5hAqyLg1I-Zv:b-nktds4) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L233)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

We are writing it for both. In future UI can also use this to manipulate delegate tags.
Do we still need to add @ApiKeyAuthorized
</blockquote></details>

Please ensure that the api works for both ui and apikey
Work with @arpit3107  to know more

**Review by mdmilic — COMMENTED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: Some of the checks are too old. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 534 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3fC673jFKReneffwv:-MwQJK8r6RioNp-tMN7y:b-ik1ykx) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L534)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

> for NG we need to add the owner (org/project) as filters in the query, but also isNg in general (so that you don't update a CG delegate by acci
I see, as of now this API will be invoked directly via customer (via script or something) so its tedious for them to send org/project ids. Let's keep this as is for now, till we make UI also to use this.
I'll add isNg in filter to avoid manipulating CG delegates.
</blockquote></details>

I resolved the comment, but I don't think that's a good idea, so up to you to decide. If we don't put the org/proj now, then when we do want to relax our uniqueness, it would be backwards incompatible change (will break customers tasks or scripts calling this API), which will make it much harder for us to introduce this change. Having said that, maybe we already have this problem elsewhere (don't know), but even if we do I think we shouldn't make it worse. 
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 552 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3iG_v5gheQ_mwcQtw:-MwQKnL34_EdMKVpYEuQ:bquwqjo) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L552)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

I am updating 'updatedUuid' here to print it as log later in this function. Not sure if streams can do that. Leaving as is.
</blockquote></details>

Resolved but just to clarify (bigger question is if you need the get from cache):
I think you probably don't need to `delegateCache.get`, in which case you don't need `streams` or `for`, just:
```
final List<String> updatedUuid = delegatesToBeUpdated.asList();
```
If you really do need a delegate cache for some reason, you could do it with `for` like you did, or alternatively with `streams`:
```
final List<String> updatedUuid = delegatesToBeUpdated.asList().stream().peek(delegate -> delegateCache.get(accountId, delegate.getUuid(), true)).collect(toList());
```

___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 553 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3j0Br144HbGHjvK5p:-MwQMON-2Ah9MT87yGz5:b-ubgl6l) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L553)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

This is to update delegatecache(s), since we updated the entries in backend.
</blockquote></details>

You invalidated the cache, so wouldn't they just get updated once someone tries to get them for real? Why do we need to force the update. 
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 556 at r2](https://reviewable.io/reviews/harness/harness-core/30778#-MwQM8jQ3K0ZOfXZSl2V:-MwQM8jR8ptFzykJijUB:b24io2g) ([raw file](https://github.com/harness/harness-core/blob/6bf51666d3ca1e396f29df9fea1b953c714bb038/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L556)):*
> ```Java
>       delegateCache.get(accountId, delegate.getUuid(), true);
>     }
>     log.info("Updating tags for delegate group: {} Delegate:{} tags:{}", delegateGroupName, updatedUuid.toString(),
> ```

no need for explicit `toString()`

**Comment by abhinav-harness** [comment]
___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 233 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3eSXwAPudhiQr696X:-MwQr3VdDbWuGfEO_Mam:boy6qy6) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L233)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

Please ensure that the api works for both ui and apikey
Work with @arpit3107  to know more
</blockquote></details>

is it ng resource?

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: Some of the checks are too old. 1 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 233 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3eSXwAPudhiQr696X:-MwQvK_69GR_s4GdBoWl:bx0kr02) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L233)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

is it ng resource?
</blockquote></details>

Yes this is only for NG
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 534 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3fC673jFKReneffwv:-MwQp7SM9ah-9NLgRhyY:blojmnv) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L534)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

I resolved the comment, but I don't think that's a good idea, so up to you to decide. If we don't put the org/proj now, then when we do want to relax our uniqueness, it would be backwards incompatible change (will break customers tasks or scripts calling this API), which will make it much harder for us to introduce this change. Having said that, maybe we already have this problem elsewhere (don't know), but even if we do I think we shouldn't make it worse. 
</blockquote></details>

I don't think so it will be hard to force these later. They'll be optional query params, which client can reject or choose to send. But in my opinion we should keeps thinks simple for direct API invocation (avoid too much IDs in API).
I am not sure future plan of UI change on this, may be @raghuatwings can decide on that.
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 552 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3iG_v5gheQ_mwcQtw:-MwQq4gf1TAAS72yoNfh:b-rrfs9s) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L552)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Resolved but just to clarify (bigger question is if you need the get from cache):
I think you probably don't need to `delegateCache.get`, in which case you don't need `streams` or `for`, just:
```
final List<String> updatedUuid = delegatesToBeUpdated.asList();
```
If you really do need a delegate cache for some reason, you could do it with `for` like you did, or alternatively with `streams`:
```
final List<String> updatedUuid = delegatesToBeUpdated.asList().stream().peek(delegate -> delegateCache.get(accountId, delegate.getUuid(), true)).collect(toList());
```

</blockquote></details>

Ohh yeah, get on cache is to ensure consistency between backend and cache. I can use streams like you mentioned.
Btw just to add "delegatesToBeUpdated.asList();" will give you lot of information on delegates (not just UUID) which we don't bother about.
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 553 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3j0Br144HbGHjvK5p:-MwQu6I52xz7j8UsFZmu:bq3bslj) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L553)):*
> final List<String> updatedUuid = delegatesToBeUpdated.asList().stream().peek(delegate -> delegateCache.get(accountId, delegate.getUuid(), true)).collect(toList());
Yes, being more proactive here, leaving cache in same state as before.. to avoid read time latency.

**Review by bharatgoelharness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: 1 change requests. 1. unresolved discussions.

**Review by bharatgoelharness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: 1 change requests. 1. unresolved discussions.

**Review by mdmilic — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: 1 change requests. 1. unresolved discussions. 

___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 534 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3fC673jFKReneffwv:-Mwhr-u500tncDI7vfCB:b-byvg7l) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L534)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

I don't think so it will be hard to force these later. They'll be optional query params, which client can reject or choose to send. But in my opinion we should keeps thinks simple for direct API invocation (avoid too much IDs in API).
I am not sure future plan of UI change on this, may be @raghuatwings can decide on that.

</blockquote></details>

They are optional, but the problem is that we are going to change the behavior which makes it backwards incompatible. So any scripts that customers have calling this API will break. That's why I think we should prevent that straight away. It's not that hard to do it. They are just two extra query params.

**Comment by abhinav-harness** [comment]
___
*[420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java, line 556 at r2](https://reviewable.io/reviews/harness/harness-core/30778#-MwQM8jQ3K0ZOfXZSl2V:-MxXdn4hB77ZpEm8wrg1:bwrjdrz) ([raw file](https://github.com/harness/harness-core/blob/6bf51666d3ca1e396f29df9fea1b953c714bb038/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L556)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

no need for explicit `toString()`
</blockquote></details>

Infact String.value of also looks redundant.

**Review by abhinav-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: 7 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java, line 233 at r1](https://reviewable.io/reviews/harness/harness-core/30778#-Mw3eSXwAPudhiQr696X:-MxXdzRM7k1yGqZtYFS0:b-me7lmt) ([raw file](https://github.com/harness/harness-core/blob/106226b71655c5f79bd70a2a9daa7eaa8f457645/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L233)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Yes this is only for NG
</blockquote></details>

@gaurav-jain1 we don't need so many annotations for manager resource -- ng-manager we need it for generating open API specs.

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 1 of 4 files at r2, 4 of 10 files at r4, 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30778)* status: 1 change requests. 1. unresolved discussions.

**Pull request merged**